### PR TITLE
nats provider and embedded server

### PR DESCRIPTION
**Pull request by kunzel-andrew**

### CODE REVIEW

**Comment by kunzel-andrew** [comment]
Current dependencies on/for this PR:
* main
  * **PR #20** <a href="https://app.graphite.dev/github/pr/shopmonkeyus/eds-server/20" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shopmonkeyus/eds-server/20?utm_source=stack-comment).

File: internal/provider/nats.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+// Stop the provider and return an error or nil if ok
+func (p *NatsProvider) Stop() error {
+	p.nc.Close()
+	return nil
+}
+
+// Process data received and return an error or nil if processed ok
+func (p *NatsProvider) Process(data datatypes.ChangeEventPayload) error {
+	p.logger.Info("Republish Message to: dbchange.%s.%s.%s",  data.GetTable(), data.GetOperation(), *data.GetLocationID())
+	subject := fmt.Sprintf("dbchange.%s.%s.%s", data.GetTable(), data.GetOperation(), *data.GetLocationID())
```
Comment (jhaynie) [inline internal/provider/nats.go:51]: you'll want to add this check:

image

File: cmd/start.go [hunk 2]
```diff
@@ -96,7 +105,26 @@ var startCmd = &cobra.Command{
...
+			}
+
+			go ns.Start()
+
+			for !ns.ReadyForConnections(4 * time.Second) {
+				logger.Info("Waiting for nats server to start...")
+			}
+			logger.Info("Nats server started at url: %s", ns.ClientURL())
+			runProvider(logger, ns.ClientURL(), dryRun, runProviderCallback)
+		}
```
Comment (jhaynie) [inline cmd/start.go:125]: put the other run in an else?

File: cmd/start.go [hunk 3]
```diff
@@ -96,7 +105,26 @@ var startCmd = &cobra.Command{
 			<-c
 			logger.Info("stopped message processor")
 			return nil
-		})
+		}
+
+		if embedNats {
+			opts := &server.Options{}
```
Comment (jhaynie) [inline cmd/start.go:111]: probably also want to expose some of the server options in case we need to change port or credentials

**Review by jhaynie — CHANGES_REQUESTED** [review]
📈

File: internal/provider/nats.go (hunk 1 above)
Comment (kunzel-andrew) [reply internal/provider/nats.go:87]: This is already done prior to the messages hitting the NatsProvider Processor so anything we receive will already only be the public models.

File: cmd/start.go (hunk 2 above)
Comment (kunzel-andrew) [reply cmd/start.go:135]: We resolved this by using a waitgroup and running the blocking provider in a go routine.

**Review by jhaynie — APPROVED** [review]

**Pull request merged**